dynamiqs/time_tensor.py

```diff
@@ -0,0 +1,257 @@
+        return self.__repr__()
+
+    def size(self, dim: int) -> int:
+        """Size along a given dimension."""
+        return self.shape[dim]
+
+    def dim(self) -> int:
+        """Get the number of dimensions."""
+        return len(self.shape)
+
+    @property
+    def ndim(self) -> int:
+        """Number of dimensions."""
+        return self.dim()
+
+
+class ConstantTimeTensor(TimeTensor):
+    def __init__(self, tensor: Tensor):
+        self.tensor = tensor
+
+    @property
+    def dtype(self) -> torch.dtype:
+        return self.tensor.dtype
+
+    @property
+    def device(self) -> torch.device:
+        return self.tensor.device
+
+    @property
+    def shape(self) -> torch.Size:
+        return self.tensor.shape
+
+    def __call__(self, t: float) -> Tensor:
+        return self.tensor
+
+    def view(self, *shape: int) -> TimeTensor:
+        return ConstantTimeTensor(self.tensor.view(*shape))
+
+    def adjoint(self) -> TimeTensor:
+        return ConstantTimeTensor(self.tensor.adjoint())
+
+    def __neg__(self) -> TimeTensor:
+        return ConstantTimeTensor(-self.tensor)
+
+    def __mul__(self, other: Number | Tensor) -> TimeTensor:
+        super().__mul__(other)
+        return ConstantTimeTensor(self.tensor * other)
+
+
+class CallableTimeTensor(TimeTensor):
+    def __init__(self, f: callable[[float], Tensor], f0: Tensor):
+        # f0 carries all the transformation on the shape
+        self.f = f
+        self.f0 = f0
+
+    @property
+    def dtype(self) -> torch.dtype:
+        return self.f0.dtype
+
+    @property
+    def device(self) -> torch.device:
+        return self.f0.device
+
+    @property
+    def shape(self) -> torch.Size:
+        return self.f0.shape
+
+    def __call__(self, t: float) -> Tensor:
+        return self.f(t).view(self.shape)
+
+    def view(self, *shape: int) -> TimeTensor:
+        f = self.f
+        f0 = self.f0.view(*shape)
+        return CallableTimeTensor(f, f0)
+
+    @abstractmethod
+    def adjoint(self) -> TimeTensor:
+        f = lambda t: self.f(t).adjoint()
+        f0 = self.f0.adjoint()
+        return CallableTimeTensor(f, f0)
+
+    def __neg__(self) -> TimeTensor:
+        f = lambda t: -self.f(t)
+        f0 = -self.f0
+        return CallableTimeTensor(f, f0)
+
+    def __mul__(self, other: Number | Tensor) -> TimeTensor:
+        super().__mul__(other)
+        f = lambda t: self.f(t) * other
+        f0 = self.f0 * other
+        return CallableTimeTensor(f, f0)
+
+
+def time_tensor_add(x: TimeTensor, y: Tensor | TimeTensor) -> TimeTensor:
+    if isinstance(x, ConstantTimeTensor):
+        if isinstance(y, Tensor):
+            return ConstantTimeTensor(x.tensor + y)
+        elif isinstance(y, ConstantTimeTensor):
+            return ConstantTimeTensor(x.tensor + y.tensor)
+        elif isinstance(y, CallableTimeTensor):
+            f = lambda t: y.f(t) + x.tensor
+            f0 = y.f0 + x.tensor
+            return CallableTimeTensor(f, f0)
```

Comment:
You are missing the default case, it should at least crash instead of silently returning `None`